From 66db038dcb41295ea86598882181cffdb5bb2781 Mon Sep 17 00:00:00 2001
From: Mariaum <me@mariaum.com>
Date: Sat, 13 Oct 2018 00:51:26 -0300
Subject: [PATCH] Mute server connecting unknown packet exception.


diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index bd3dc40..29ce4be 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -52,6 +52,7 @@ import net.md_5.bungee.protocol.packet.ScoreboardScore;
 import net.md_5.bungee.protocol.packet.SetCompression;
 import net.md_5.bungee.protocol.packet.ViewDistance;
 import net.md_5.bungee.util.BufUtil;
+import net.md_5.bungee.util.MuteException; // Zartema
 import net.md_5.bungee.util.QuietException;
 
 @RequiredArgsConstructor
@@ -158,7 +159,7 @@ public class ServerConnector extends PacketHandler
     {
         if ( packet.packet == null )
         {
-            throw new QuietException( "Unexpected packet received during server login process!\n" + BufUtil.dump( packet.buf, 16 ) );
+            throw MuteException.INSTANCE; // Zartema - mute exception
         }
     }
 
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
index a66099f..5d9b47f 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
@@ -16,6 +16,7 @@ import net.md_5.bungee.connection.PingHandler;
 import net.md_5.bungee.protocol.BadPacketException;
 import net.md_5.bungee.protocol.OverflowPacketException;
 import net.md_5.bungee.protocol.PacketWrapper;
+import net.md_5.bungee.util.MuteException; // Zartema
 import net.md_5.bungee.util.QuietException;
 
 /**
@@ -154,6 +155,7 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
                     {
                         handler, cause
                     } );
+                } else if ( cause instanceof MuteException ) { // Zartema
                 } else
                 {
                     ProxyServer.getInstance().getLogger().log( Level.SEVERE, handler + " - encountered exception", cause );
diff --git a/proxy/src/main/java/net/md_5/bungee/util/MuteException.java b/proxy/src/main/java/net/md_5/bungee/util/MuteException.java
new file mode 100644
index 0000000..f7ac466
--- /dev/null
+++ b/proxy/src/main/java/net/md_5/bungee/util/MuteException.java
@@ -0,0 +1,27 @@
+package net.md_5.bungee.util;
+
+/**
+ * Exception without a stack trace component and a message.
+ */
+public class MuteException extends RuntimeException
+{
+
+    public static final MuteException INSTANCE = new MuteException();
+
+    public MuteException()
+    {
+        super();
+    }
+
+    @Override
+    public Throwable initCause(Throwable cause)
+    {
+        return this;
+    }
+
+    @Override
+    public Throwable fillInStackTrace()
+    {
+        return this;
+    }
+}
-- 
2.7.4

