From e8fd264b5773784f43ae17698d6c095659d3317d Mon Sep 17 00:00:00 2001
From: Mariaum <me@mariaum.com>
Date: Mon, 25 May 2020 13:31:31 -0300
Subject: [PATCH] Fix keepalive queue causing issues.

See https://github.com/SpigotMC/BungeeCord/pull/2786

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 359c90e0..96645f37 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -130,10 +130,11 @@ public class UpstreamBridge extends PacketHandler
     @Override
     public void handle(KeepAlive alive) throws Exception
     {
-        KeepAliveData keepAliveData = con.getServer().getKeepAlives().poll();
+        KeepAliveData keepAliveData = con.getServer().getKeepAlives().peek();
 
         if ( keepAliveData != null && alive.getRandomId() == keepAliveData.getId() )
         {
+            con.getServer().getKeepAlives().remove( keepAliveData );
             int newPing = (int) ( System.currentTimeMillis() - keepAliveData.getTime() );
             con.getTabListHandler().onPingChange( newPing );
             con.setPing( newPing );
-- 
2.21.0.windows.1

