From 620bff8b2e6b9390bc46e1b56cfeff8591643ab2 Mon Sep 17 00:00:00 2001
From: Mariaum <me@mariaum.com>
Date: Mon, 30 Sep 2019 21:33:27 -0300
Subject: [PATCH] Temporary optimization to empty packets until Travertine is
 updated


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
index 01997098..f1da62a6 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
@@ -46,7 +46,7 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
         {
             // Waterfall start
             if (in.readableBytes() == 0) {
-                if (!allowEmptyPackets) throw new BadPacketException("Empty minecraft packet!");
+                // if (!allowEmptyPackets) throw new BadPacketException("Empty minecraft packet!");
                 return;
             }
             // Waterfall end
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
index d2e319f3..c138c4f4 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Varint21FrameDecoder.java
@@ -41,7 +41,7 @@ public class Varint21FrameDecoder extends ByteToMessageDecoder
                 in.resetReaderIndex();
                 int length = DefinedPacket.readVarInt( in );
                 // Waterfall end
-                if ( length == 0 && !allowEmptyPackets) // Waterfall
+                if ( false && length == 0 && !allowEmptyPackets) // Waterfall
                 {
                     // Travertine start - vanilla 1.7 client sometimes sends empty packets.
                     long currentTime = System.currentTimeMillis();
-- 
2.17.1

