From 4f50b291a30426700b069864c098d3e1b5fd9035 Mon Sep 17 00:00:00 2001
From: Mariaum <me@mariaum.com>
Date: Fri, 16 Oct 2020 00:38:17 -0300
Subject: [PATCH] Allow to hide sensitive command args from logs.


diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/Command.java b/api/src/main/java/net/md_5/bungee/api/plugin/Command.java
index b4566327..71a5c085 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/Command.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/Command.java
@@ -66,4 +66,16 @@ public abstract class Command
     {
         return permission == null || permission.isEmpty() || sender.hasPermission( permission );
     }
+
+    // Zartema start - allow to hide sensitive args from logging
+    /**
+     * Check if the command has sensitive arguments like passwords or tokens
+     *
+     * @return whether this commands's arguments are sensitive
+     */
+    public boolean isSensitiveArgs()
+    {
+        return false;
+    }
+    // Zartema end
 }
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index d2cb63a4..79aaa896 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -225,7 +225,7 @@ public final class PluginManager
                 {
                     proxy.getLogger().log( Level.INFO, "{0} executed command: /{1}", new Object[]
                     {
-                        sender.getName(), commandLine
+                        sender.getName(), command.isSensitiveArgs() ? command.getName() : commandLine
                     } );
                 }
                 command.execute( sender, args );
-- 
2.34.1

