From 0a6774adadc0b07e78fcc547ffd60b79ed22c3f8 Mon Sep 17 00:00:00 2001
From: Mariaum <me@mariaum.com>
Date: Thu, 5 Nov 2020 21:33:37 -0300
Subject: [PATCH] Don't disable connection throttle for every listener when
 using proxy protocol.


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 545475a42..e1b0ae209 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -361,12 +361,12 @@ public class BungeeCord extends ProxyServer
             if ( info.isProxyProtocol() )
             {
                 getLogger().log( Level.WARNING, "Using PROXY protocol for listener {0}, please ensure this listener is adequately firewalled.", info.getSocketAddress() );
-
+                /* Zartema - Don't disable connection throttle for every listener when using proxy protocol.
                 if ( connectionThrottle != null )
                 {
                     connectionThrottle = null;
                     getLogger().log( Level.WARNING, "Since PROXY protocol is in use, internal connection throttle has been disabled." );
-                }
+                }*/
             }
 
             ChannelFutureListener listener = new ChannelFutureListener()
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index b318deab9..6747c2d57 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -65,13 +65,14 @@ public class PipelineUtils
         ProxyServer.getInstance().unsafe().setFrontendChannelInitializer( BungeeChannelInitializer.create( ch ->
         {
             SocketAddress remoteAddress = ( ch.remoteAddress() == null ) ? ch.parent().localAddress() : ch.remoteAddress();
+            // Zartema start - Don't disable connection throttle for every listener when using proxy protocol.
+            ListenerInfo listener = ch.attr( LISTENER ).get();
 
-            if ( BungeeCord.getInstance().getConnectionThrottle() != null && BungeeCord.getInstance().getConnectionThrottle().throttle( remoteAddress ) )
+            if ( !listener.isProxyProtocol() && BungeeCord.getInstance().getConnectionThrottle() != null && BungeeCord.getInstance().getConnectionThrottle().throttle( remoteAddress ) )
             {
                 return false;
             }
-
-            ListenerInfo listener = ch.attr( LISTENER ).get();
+            // Zartema end
 
             if ( BungeeCord.getInstance().getPluginManager().callEvent( new ClientConnectEvent( remoteAddress, listener ) ).isCancelled() )
             {
-- 
2.34.1

