From a402391b6605bda384f46274b80ad110faf4dc43 Mon Sep 17 00:00:00 2001
From: mariaum <me@mariaum.com>
Date: Sat, 14 Oct 2023 02:53:26 -0300
Subject: [PATCH] Fix sendPacketQueued() throwing NPEs.


diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnection.java b/proxy/src/main/java/net/md_5/bungee/ServerConnection.java
index 5d5404502..b1c91bd73 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnection.java
@@ -45,6 +45,7 @@ public class ServerConnection implements Server
 
     public void sendPacketQueued(DefinedPacket packet)
     {
+        if (ch.isClosed()) return; // Zartema
         Protocol encodeProtocol = ch.getEncodeProtocol();
         if ( !encodeProtocol.TO_SERVER.hasPacket( packet.getClass(), ch.getEncodeVersion() ) )
         {
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 43230fbdf..ffdc9c8a4 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -192,6 +192,7 @@ public final class UserConnection implements ProxiedPlayer
 
     public void sendPacketQueued(DefinedPacket packet)
     {
+        if (ch.isClosed()) return; // Zartema
         Protocol encodeProtocol = ch.getEncodeProtocol();
         if ( !encodeProtocol.TO_CLIENT.hasPacket( packet.getClass(), getPendingConnection().getVersion() ) )
         {
-- 
2.40.1.windows.1

