From be6ae4b6b74746dd902e079c07f8ba7a3ac3694d Mon Sep 17 00:00:00 2001
From: mariaum <me@mariaum.com>
Date: Sat, 27 Apr 2024 04:08:32 -0300
Subject: [PATCH] Allow spoofing chat in 1.19+.

Should work with ViaVersion at least.

diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 71b4288c6..51b0c46f5 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -491,7 +491,8 @@ public final class UserConnection implements ProxiedPlayer
         Preconditions.checkState( server != null, "Not connected to server" );
         if ( getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_19 )
         {
-            throw new UnsupportedOperationException( "Cannot spoof chat on this client version!" );
+            server.getCh().write( new net.md_5.bungee.protocol.packet.ClientChat( message, 0, 0, new byte[256], false, new net.md_5.bungee.protocol.ChatChain( com.google.common.collect.ImmutableList.of(), com.google.common.collect.ImmutableList.of() ), new net.md_5.bungee.protocol.SeenMessages( 0, new java.util.BitSet() ) ) ); // Zartema - Allow spoofing chat in 1.19+.
+            return;
         }
         server.getCh().write( new Chat( message ) );
     }
-- 
2.34.1

