From 4f6df564ecbe07c230aee78002c9070692f04a42 Mon Sep 17 00:00:00 2001
From: mariaum <me@mariaum.com>
Date: Sat, 27 Apr 2024 04:08:32 -0300
Subject: [PATCH] Allow spoofing chat in 1.19+.

Should work with ViaVersion at least.

diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 60f506c7..257f4e4a 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -509,7 +509,8 @@ public final class UserConnection implements ProxiedPlayer
         Preconditions.checkState( server != null, "Not connected to server" );
         if ( getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_19 )
         {
-            throw new UnsupportedOperationException( "Cannot spoof chat on this client version!" );
+            server.getCh().write( new net.md_5.bungee.protocol.packet.ClientChat( message, 0, 0, new byte[256], false, new net.md_5.bungee.protocol.ChatChain( com.google.common.collect.ImmutableList.of(), com.google.common.collect.ImmutableList.of() ), new net.md_5.bungee.protocol.SeenMessages( 0, new java.util.BitSet() ), (byte) 0 ) ); // Zartema - Allow spoofing chat in 1.19+.
+            return;
         }
         server.getCh().write( new Chat( message ) );
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index ec8c7a3f..f652b672 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -194,19 +194,54 @@ public class UpstreamBridge extends PacketHandler
     @Override
     public void handle(ClientChat chat) throws Exception
     {
-        handleChat( chat.getMessage() );
+        // Zartema start - Allow spoofing chat in 1.19+.
+        String message = handleChat( chat.getMessage() );
+        if ( !java.util.Objects.equals( message, chat.getMessage() ) )
+        {
+            if ( message != null ) {
+                con.getServer().unsafe().sendPacket( new net.md_5.bungee.protocol.packet.ClientChat( message, 0, 0, new byte[256], false, new net.md_5.bungee.protocol.ChatChain( com.google.common.collect.ImmutableList.of(), com.google.common.collect.ImmutableList.of() ), new net.md_5.bungee.protocol.SeenMessages( 0, new java.util.BitSet() ), (byte) 0 ) );
+            }
+            throw CancelSendSignal.INSTANCE;
+        }
+        // Zartema end
     }
 
     @Override
     public void handle(ClientCommand command) throws Exception
     {
-        handleChat( "/" + command.getCommand(), command );
+        // Zartema start - Allow spoofing chat in 1.19+.
+        String message = handleChat( "/" + command.getCommand(), command );
+        if ( !java.util.Objects.equals( message, "/" + command.getCommand() ) )
+        {
+            if ( message != null ) {
+                if ( message.startsWith("/") ) {
+                    con.getServer().unsafe().sendPacket( new net.md_5.bungee.protocol.packet.UnsignedClientCommand( message.substring(1) ) );
+                } else {
+                    con.getServer().unsafe().sendPacket( new net.md_5.bungee.protocol.packet.ClientChat( message, 0, 0, new byte[256], false, new net.md_5.bungee.protocol.ChatChain( com.google.common.collect.ImmutableList.of(), com.google.common.collect.ImmutableList.of() ), new net.md_5.bungee.protocol.SeenMessages( 0, new java.util.BitSet() ), (byte) 0 ) );
+                }
+            }
+            throw CancelSendSignal.INSTANCE;
+        }
+        // Zartema end
     }
 
     @Override
     public void handle(UnsignedClientCommand command) throws Exception
     {
-        handleChat( "/" + command.getCommand(), null); // Waterfall
+        // Zartema start - Allow spoofing chat in 1.19+.
+        String message = handleChat( "/" + command.getCommand(), null );
+        if ( !java.util.Objects.equals( message, "/" + command.getCommand() ) )
+        {
+            if ( message != null ) {
+                if ( message.startsWith("/") ) {
+                    con.getServer().unsafe().sendPacket( new net.md_5.bungee.protocol.packet.UnsignedClientCommand( message.substring(1) ) );
+                } else {
+                    con.getServer().unsafe().sendPacket( new net.md_5.bungee.protocol.packet.ClientChat( message, 0, 0, new byte[256], false, new net.md_5.bungee.protocol.ChatChain( com.google.common.collect.ImmutableList.of(), com.google.common.collect.ImmutableList.of() ), new net.md_5.bungee.protocol.SeenMessages( 0, new java.util.BitSet() ), (byte) 0 ) );
+                }
+            }
+            throw CancelSendSignal.INSTANCE;
+        }
+        // Zartema end
     }
 
     private String handleChat(String message)
-- 
2.51.0.windows.1

