From 82b4f05b4ff1e0e4202839455ac9adfa9f98049b Mon Sep 17 00:00:00 2001
From: mariaum <me@mariaum.com>
Date: Tue, 15 Apr 2025 17:51:30 -0300
Subject: [PATCH] Fix disable tab list rewrite config not fully disabling
 rewrites.


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 5255496f..90772ae2 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -195,15 +195,27 @@ public class DownstreamBridge extends PacketHandler
     @Override
     public void handle(PlayerListItemRemove playerList) throws Exception
     {
-        con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
-        throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
+        // Zartema start
+        boolean skipRewrites = bungee.getConfig().isDisableTabListRewrite();
+        con.getTabListHandler().onUpdate( skipRewrites ? playerList : TabList.rewrite( playerList ) );
+        if ( !skipRewrites )
+        {
+            throw CancelSendSignal.INSTANCE; // Only throw if profile rewriting is enabled
+        }
+        // Zartema end
     }
 
     @Override
     public void handle(PlayerListItemUpdate playerList) throws Exception
     {
-        con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
-        throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
+        // Zartema start
+        boolean skipRewrites = bungee.getConfig().isDisableTabListRewrite();
+        con.getTabListHandler().onUpdate( skipRewrites ? playerList : TabList.rewrite( playerList ) );
+        if ( !skipRewrites )
+        {
+            throw CancelSendSignal.INSTANCE; // Only throw if profile rewriting is enabled
+        }
+        // Zartema end
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 00e7015c..33afca71 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -83,38 +83,42 @@ public class UpstreamBridge extends PacketHandler
 
         if ( con.getServer() != null )
         {
-            // Manually remove from everyone's tab list
-            // since the packet from the server arrives
-            // too late
-            // TODO: This should only done with server_unique
-            //       tab list (which is the only one supported
-            //       currently)
-            PlayerListItem oldPacket = new PlayerListItem();
-            oldPacket.setAction( PlayerListItem.Action.REMOVE_PLAYER );
-            PlayerListItem.Item item = new PlayerListItem.Item();
-            item.setUuid( con.getRewriteId() );
-            oldPacket.setItems( new PlayerListItem.Item[]
-            {
-                item
-            } );
-
-            PlayerListItemRemove newPacket = new PlayerListItemRemove();
-            newPacket.setUuids( new UUID[]
-            {
-                con.getRewriteId()
-            } );
-
-            for ( ProxiedPlayer player : con.getServer().getInfo().getPlayers() )
-            {
-                if ( player.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_19_3 )
-                {
-                    // need to queue, because players in config state could receive it
-                    player.unsafe().sendPacketQueued( newPacket );
-                } else if ( ProtocolConstants.isAfterOrEq( player.getPendingConnection().getVersion(), ProtocolConstants.MINECRAFT_1_8 ) ) // Travertine
+            // Zartema start
+            if (!bungee.getConfig().isDisableTabListRewrite()) {
+                // Manually remove from everyone's tab list
+                // since the packet from the server arrives
+                // too late
+                // TODO: This should only done with server_unique
+                //       tab list (which is the only one supported
+                //       currently)
+                PlayerListItem oldPacket = new PlayerListItem();
+                oldPacket.setAction( PlayerListItem.Action.REMOVE_PLAYER );
+                PlayerListItem.Item item = new PlayerListItem.Item();
+                item.setUuid( con.getRewriteId() );
+                oldPacket.setItems( new PlayerListItem.Item[]
+                        {
+                                item
+                        } );
+
+                PlayerListItemRemove newPacket = new PlayerListItemRemove();
+                newPacket.setUuids( new UUID[]
+                        {
+                                con.getRewriteId()
+                        } );
+
+                for ( ProxiedPlayer player : con.getServer().getInfo().getPlayers() )
                 {
-                    player.unsafe().sendPacket( oldPacket );
+                    if ( player.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_19_3 )
+                    {
+                        // need to queue, because players in config state could receive it
+                        player.unsafe().sendPacketQueued( newPacket );
+                    } else if ( ProtocolConstants.isAfterOrEq( player.getPendingConnection().getVersion(), ProtocolConstants.MINECRAFT_1_8 ) ) // Travertine
+                    {
+                        player.unsafe().sendPacket( oldPacket );
+                    }
                 }
             }
+            // Zartema end
             con.getServer().disconnect( "Quitting" );
         }
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
index e30492cb..80da6f8f 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
@@ -120,7 +120,7 @@ class EntityMap_1_10 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -162,7 +162,7 @@ class EntityMap_1_10 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
index 42c6d56f..dd2ab0cc 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
@@ -120,7 +120,7 @@ class EntityMap_1_11 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -163,7 +163,7 @@ class EntityMap_1_11 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
index 811bcac8..2dfc2ce7 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
@@ -120,7 +120,7 @@ class EntityMap_1_12 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -163,7 +163,7 @@ class EntityMap_1_12 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
index e8c38c6e..f6ef338c 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
@@ -120,7 +120,7 @@ class EntityMap_1_12_1 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -163,7 +163,7 @@ class EntityMap_1_12_1 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
index 5fa12256..22b4a435 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
@@ -120,7 +120,7 @@ class EntityMap_1_13 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -163,7 +163,7 @@ class EntityMap_1_13 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x28 /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x28 /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
index 33f25c92..07c710fe 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
@@ -119,7 +119,7 @@ class EntityMap_1_14 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -167,7 +167,7 @@ class EntityMap_1_14 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
index 60cdf04c..3cfc0eda 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
@@ -119,7 +119,7 @@ class EntityMap_1_15 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -167,7 +167,7 @@ class EntityMap_1_15 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
index 735bd6df..e52578f2 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
@@ -119,7 +119,7 @@ class EntityMap_1_16 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -167,7 +167,7 @@ class EntityMap_1_16 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2C /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x2C /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java
index 9560ff3d..8189516e 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java
@@ -41,7 +41,7 @@ class EntityMap_1_16_2 extends EntityMap
             int idLength = packet.readerIndex() - readerIndex - packetIdLength;
             UUID uuid = DefinedPacket.readUUID( packet );
             UserConnection player;
-            if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+            if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
             {
                 int previous = packet.writerIndex();
                 packet.readerIndex( readerIndex );
@@ -61,7 +61,7 @@ class EntityMap_1_16_2 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == spectateId )
+        if ( packetId == spectateId && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
index d3b52da4..bfca1ad8 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
@@ -120,7 +120,7 @@ class EntityMap_1_8 extends EntityMap
             int idLength = packet.readerIndex() - readerIndex - packetIdLength;
             UUID uuid = DefinedPacket.readUUID( packet );
             UserConnection player;
-            if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+            if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
             {
                 int previous = packet.writerIndex();
                 packet.readerIndex( readerIndex );
@@ -156,7 +156,7 @@ class EntityMap_1_8 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x18 /* Spectate */ )
+        if ( packetId == 0x18 /* Spectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
index 0cd0bf06..37d9dc99 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
@@ -120,7 +120,7 @@ class EntityMap_1_9 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -162,7 +162,7 @@ class EntityMap_1_9 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
index 930a7b5a..271f53f8 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
@@ -120,7 +120,7 @@ class EntityMap_1_9_4 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -162,7 +162,7 @@ class EntityMap_1_9_4 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
-- 
2.34.1

