From e0e190e33a25d6091cde37a2a25e14698e4ec729 Mon Sep 17 00:00:00 2001
From: mariaum <me@mariaum.com>
Date: Tue, 15 Apr 2025 17:51:30 -0300
Subject: [PATCH] Fix disable tab list rewrite config not fully disabling
 rewrites.


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 1f117c83..f8841670 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -194,15 +194,27 @@ public class DownstreamBridge extends PacketHandler
     @Override
     public void handle(PlayerListItemRemove playerList) throws Exception
     {
-        con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
-        throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
+        // Zartema start
+        boolean skipRewrites = bungee.getConfig().isDisableTabListRewrite();
+        con.getTabListHandler().onUpdate( skipRewrites ? playerList : TabList.rewrite( playerList ) );
+        if ( !skipRewrites )
+        {
+            throw CancelSendSignal.INSTANCE; // Only throw if profile rewriting is enabled
+        }
+        // Zartema end
     }
 
     @Override
     public void handle(PlayerListItemUpdate playerList) throws Exception
     {
-        con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
-        throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
+        // Zartema start
+        boolean skipRewrites = bungee.getConfig().isDisableTabListRewrite();
+        con.getTabListHandler().onUpdate( skipRewrites ? playerList : TabList.rewrite( playerList ) );
+        if ( !skipRewrites )
+        {
+            throw CancelSendSignal.INSTANCE; // Only throw if profile rewriting is enabled
+        }
+        // Zartema end
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index d3b4c646..83788c10 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -81,38 +81,42 @@ public class UpstreamBridge extends PacketHandler
 
         if ( con.getServer() != null )
         {
-            // Manually remove from everyone's tab list
-            // since the packet from the server arrives
-            // too late
-            // TODO: This should only done with server_unique
-            //       tab list (which is the only one supported
-            //       currently)
-            PlayerListItem oldPacket = new PlayerListItem();
-            oldPacket.setAction( PlayerListItem.Action.REMOVE_PLAYER );
-            PlayerListItem.Item item = new PlayerListItem.Item();
-            item.setUuid( con.getRewriteId() );
-            oldPacket.setItems( new PlayerListItem.Item[]
-            {
-                item
-            } );
-
-            PlayerListItemRemove newPacket = new PlayerListItemRemove();
-            newPacket.setUuids( new UUID[]
-            {
-                con.getRewriteId()
-            } );
-
-            for ( ProxiedPlayer player : con.getServer().getInfo().getPlayers() )
-            {
-                if ( player.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_19_3 )
-                {
-                    // need to queue, because players in config state could receive it
-                    player.unsafe().sendPacketQueued( newPacket );
-                } else if ( ProtocolConstants.isAfterOrEq( player.getPendingConnection().getVersion(), ProtocolConstants.MINECRAFT_1_8 ) ) // Travertine
+            // Zartema start
+            if (!bungee.getConfig().isDisableTabListRewrite()) {
+                // Manually remove from everyone's tab list
+                // since the packet from the server arrives
+                // too late
+                // TODO: This should only done with server_unique
+                //       tab list (which is the only one supported
+                //       currently)
+                PlayerListItem oldPacket = new PlayerListItem();
+                oldPacket.setAction( PlayerListItem.Action.REMOVE_PLAYER );
+                PlayerListItem.Item item = new PlayerListItem.Item();
+                item.setUuid( con.getRewriteId() );
+                oldPacket.setItems( new PlayerListItem.Item[]
+                        {
+                                item
+                        } );
+
+                PlayerListItemRemove newPacket = new PlayerListItemRemove();
+                newPacket.setUuids( new UUID[]
+                        {
+                                con.getRewriteId()
+                        } );
+
+                for ( ProxiedPlayer player : con.getServer().getInfo().getPlayers() )
                 {
-                    player.unsafe().sendPacket( oldPacket );
+                    if ( player.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_19_3 )
+                    {
+                        // need to queue, because players in config state could receive it
+                        player.unsafe().sendPacketQueued( newPacket );
+                    } else if ( ProtocolConstants.isAfterOrEq( player.getPendingConnection().getVersion(), ProtocolConstants.MINECRAFT_1_8 ) ) // Travertine
+                    {
+                        player.unsafe().sendPacket( oldPacket );
+                    }
                 }
             }
+            // Zartema end
             con.getServer().disconnect( "Quitting" );
         }
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
index f42dc722..a0bf1401 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_10.java
@@ -122,7 +122,7 @@ class EntityMap_1_10 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -164,7 +164,7 @@ class EntityMap_1_10 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
index dc711f34..2b1ea354 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_11.java
@@ -122,7 +122,7 @@ class EntityMap_1_11 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -165,7 +165,7 @@ class EntityMap_1_11 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
index c15ab1a1..462fc2c3 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12.java
@@ -122,7 +122,7 @@ class EntityMap_1_12 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -165,7 +165,7 @@ class EntityMap_1_12 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
index c3e6f8c7..38e09789 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_12_1.java
@@ -122,7 +122,7 @@ class EntityMap_1_12_1 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -165,7 +165,7 @@ class EntityMap_1_12_1 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1E /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
index c653d664..f6210391 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_13.java
@@ -122,7 +122,7 @@ class EntityMap_1_13 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -165,7 +165,7 @@ class EntityMap_1_13 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x28 /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x28 /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
index e9b28f0d..6497107b 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_14.java
@@ -121,7 +121,7 @@ class EntityMap_1_14 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -169,7 +169,7 @@ class EntityMap_1_14 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
index eafca587..97b49bd2 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_15.java
@@ -121,7 +121,7 @@ class EntityMap_1_15 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -169,7 +169,7 @@ class EntityMap_1_15 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x2B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
index 98b1a155..71154810 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
@@ -121,7 +121,7 @@ class EntityMap_1_16 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -169,7 +169,7 @@ class EntityMap_1_16 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x2C /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x2C /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java
index ce96ac9e..23b6a8d2 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16_2.java
@@ -49,7 +49,7 @@ class EntityMap_1_16_2 extends EntityMap
             int idLength = packet.readerIndex() - readerIndex - packetIdLength;
             UUID uuid = DefinedPacket.readUUID( packet );
             UserConnection player;
-            if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+            if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
             {
                 int previous = packet.writerIndex();
                 packet.readerIndex( readerIndex );
@@ -69,7 +69,7 @@ class EntityMap_1_16_2 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == spectateId )
+        if ( packetId == spectateId && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
index 42a53153..49df1e02 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_8.java
@@ -122,7 +122,7 @@ class EntityMap_1_8 extends EntityMap
             int idLength = packet.readerIndex() - readerIndex - packetIdLength;
             UUID uuid = DefinedPacket.readUUID( packet );
             UserConnection player;
-            if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+            if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
             {
                 int previous = packet.writerIndex();
                 packet.readerIndex( readerIndex );
@@ -158,7 +158,7 @@ class EntityMap_1_8 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x18 /* Spectate */ )
+        if ( packetId == 0x18 /* Spectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
index 657e5846..52313483 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9.java
@@ -122,7 +122,7 @@ class EntityMap_1_9 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -164,7 +164,7 @@ class EntityMap_1_9 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
index 11d8e0e3..6d5c8d99 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_9_4.java
@@ -122,7 +122,7 @@ class EntityMap_1_9_4 extends EntityMap
                 int idLength = packet.readerIndex() - readerIndex - packetIdLength;
                 UUID uuid = DefinedPacket.readUUID( packet );
                 UserConnection player;
-                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                if ( !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() && ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null ) // Zartema
                 {
                     int previous = packet.writerIndex();
                     packet.readerIndex( readerIndex );
@@ -164,7 +164,7 @@ class EntityMap_1_9_4 extends EntityMap
         int packetId = DefinedPacket.readVarInt( packet );
         int packetIdLength = packet.readerIndex() - readerIndex;
 
-        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ )
+        if ( packetId == 0x1B /* Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isDisableTabListRewrite() ) // Zartema
         {
             UUID uuid = DefinedPacket.readUUID( packet );
             ProxiedPlayer player;
-- 
2.45.1.windows.1

